applications/Subcoin.md

```diff
@@ -12,7 +12,7 @@ This project aims to develop a prototype of a Bitcoin full node in Rust using th
 
 **Relevance to Substrate & Polkadot**
 
-This project showcases the versatility of Substrate as a leading blockchain development framework. With Substrate, small teams and even individuals can create production-level blockchain software efficiently. Developers within the Substrate/Polkadot ecosystem, particularly those interested in Bitcoin-related applications, stand to benefit from the tools and insights produced by this project. 
+This project showcases the versatility of Substrate as a leading blockchain development framework. With Substrate, small teams and even individuals can create production-level blockchain software efficiently. Developers within the Substrate/Polkadot ecosystem, particularly those interested in Bitcoin-related applications, stand to benefit from the tools and insights produced by this project. Specifically, the Bitcoin bridge projects can integrate the subcoin-network into their nodes to talk to the Bitcoin network directly instead of relying on the external relayers which are suffered from the single point of failure practically.
```

Comment:
>Could you add more details on how a parachain is going to be able to interact with the bitcoin network using your client?

@PieWol It's already mentioned here. By integrating the crate `subcoin-network` into their node, the projects can interact with the Bitcoin p2p network. The use of `subcoin-network` (how to integrate this crate), the exposed interfaces specifically, will be detailed in this crate's inline docs (part of the deliverable).

Reply:
@liuchengxu thanks for the speedy reply. Could you comment how that communication would look like, i.e. what protocol and data format will be used here (e.g. JSON via RPC)?

Reply:
It's the standard Bitcoin p2p protocol, see more details at https://developer.bitcoin.org/reference/p2p_networking.html

Reply:
ok, so [RPC](https://developer.bitcoin.org/reference/rpc/index.html), right?

Reply:
No, checkout https://bitcoin.stackexchange.com/questions/36113/whats-the-difference-between-the-bitcoins-peer-to-peer-protocol-and-the-rpc-ap for the differences between p2p and RPC.

Technically, `subcoin-network` will implement the Bitcoin p2p protocol, which is the key of this application. `subcoin-rpc` will implement the Bitcoin Transaction Broadcasting RPC as a deliverable in M3 added in the last commit, but not all RPCs in https://developer.bitcoin.org/reference/rpc/index.html will be developed in this application.

Reply:
@liuchengxu thanks for elaborating. So from what I've understood, the client will be capable of communicating in _Bitcoin p2p_ and provide a limited _RPC_ api. However, I still didn't quite understand how the runtime on a node would interact/communicate with the bitcoin client. Could you elaborate on that as well?

Reply:
>I still didn't quite understand how the runtime on a node would interact/communicate with the bitcoin client. 

You may be lost if the bitcoin client you mentioned here refers to another bitcoin full node. Simply put, a typical Substrate node consists of the runtime and the necessary client components. The (Substrate) runtime is in a sandbox environment and can only interact with the Substrate client, the runtime can not directly communicate with the bitcoin client. The fact is the crate `subcoin-network` (not part of the runtime) will talk to another Bitcoin client to sync the Bitcoin blocks, and then these synced blocks will be processed by the runtime.

There is quite some technical background and there are some concepts we need to align here, if you want to comprehend these things, I suggest we schedule another meeting for efficiency. Regarding the technical considerations, you may resort to a senior parity engineer to review the soundness of this application and the design notes.

Reply:
>The part I don't understand is how the runtime is going to access these synced blocks.

The runtime itself cannot proactively access synced blocks. Instead, you need to interact with the runtime by invoking the runtime API on the client side. Here's how it works:
- Receiving Blocks: Blocks downloaded from the `subcoin-network` are sent to the import queue.
- Execution: The blocks are executed using the runtime API's `execute_block()` method.
- Importing: Upon successful execution, the blocks are imported into the node's database.

If you still have questions on this point, we can chat offline sometime tomorrow. That would be more efficient.